lib/internal/process/write-coverage.js

```diff
@@ -0,0 +1,46 @@
+'use strict';
+const process = require('process');
+const path = require('path');
+const fs = require('fs');
+const mkdirSync = fs.mkdirSync;
+const writeFileSync = fs.writeFileSync;
+
+var isWritingCoverage = false;
+function writeCoverage() {
+  if (isWritingCoverage || !global.__coverage__) {
+    return;
+  }
+  isWritingCoverage = true;
+
+  const dirname = path.join(path.dirname(process.execPath), '.coverage');
+  const filename = `coverage-${process.pid}-${Date.now()}.json`;
+  try {
+    mkdirSync(dirname);
+  } catch (err) {
+    if (err.code !== 'EEXIST') {
+      console.error(err);
+      return;
+    }
+  }
+
+  const target = path.join(dirname, filename);
+  const coverageInfo = JSON.stringify(global.__coverage__);
+  try {
+    writeFileSync(target, coverageInfo);
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+function setup() {
+  var reallyReallyExit = process.reallyExit;
+
+  process.reallyExit = function(code) {
+    writeCoverage();
+    reallyReallyExit(code);
+  };
+
+  process.on('exit', writeCoverage);
+}
```

Comment:
It might be worth running the contents of `setup()` only conditionally, e.g. when `global.__coverage__` is set (I think that test should work)?